mkt/langpacks/tests/test_models.py

```diff
@@ -98,6 +108,88 @@ def test_upload_language_validation(self, get_json_data_mock):
+            'languages-provided': {
+                'es': {}
+            },
+            'languages-target': {
+                'app://*.gaiamobile.org/manifest.webapp': '2.2'
+            },
+        }
+        expected = [u'Your language pack should contain a version.']
+        with self.assertRaises(ValidationError) as e:
+            LangPack.from_upload(upload)
+        eq_(e.exception.messages, expected)
+
+    @patch('mkt.files.utils.WebAppParser.get_json_data')
+    def test_upload_existing_same_version(self, get_json_data_mock):
+        upload = self.upload('langpack')
+        langpack = self.create_langpack()
+        get_json_data_mock.return_value = {
+            'role': 'langpack',
+            'languages-provided': {
+                'es': {}
+            },
+            'languages-target': {
+                'app://*.gaiamobile.org/manifest.webapp': '2.2'
+            },
+            'version': '1.0'
+        }
+        # Works once.
+        ok_(LangPack.from_upload(upload, instance=langpack))
+
+        # Doesn't work twice, since we are re-uploading the same version.
+        expected = [u'Your language pack version must be different to the one '
+                    u'you are replacing.']
+        with self.assertRaises(ValidationError) as e:
+            LangPack.from_upload(upload, instance=langpack)
+        eq_(e.exception.messages, expected)
+
+    @patch('mkt.langpacks.models.sign_app')
+    def test_upload_sign(self, sign_app_mock):
+        eq_(LangPack.objects.count(), 0)
+        upload = self.upload('langpack')
+        langpack = LangPack.from_upload(upload)
+        ok_(langpack.pk)
+        ok_(langpack.file_version)
+        ok_(langpack.file_path)
+        eq_(LangPack.objects.count(), 1)
+        expected_args = (
+            upload.path,
+            langpack.file_path,
+            json.dumps({'id': langpack.pk, 'version': langpack.file_version})
+        )
+        sign_app_mock.assert_called_once_with(*expected_args)
+
+    @patch('mkt.langpacks.models.sign_app')
+    def test_upload_sign_existing(self, sign_app_mock):
+        langpack = self.create_langpack()
+        eq_(LangPack.objects.count(), 1)
+        upload = self.upload('langpack')
+        langpack = LangPack.from_upload(upload, instance=langpack)
+        ok_(langpack.pk)
+        ok_(langpack.file_version)
+        ok_(langpack.file_path)
+        eq_(LangPack.objects.count(), 1)
+        expected_args = (
+            upload.path,
+            langpack.file_path,
+            json.dumps({'id': langpack.pk, 'version': langpack.file_version})
+        )
+        sign_app_mock.assert_called_once_with(*expected_args)
+
+    @patch('mkt.langpacks.models.sign_app')
+    def test_upload_sign_error(self, sign_app_mock):
+        sign_app_mock.side_effect = SigningError
+        eq_(LangPack.objects.count(), 0)
+        upload = self.upload('langpack')
+        with self.assertRaises(SigningError):
+            LangPack.from_upload(upload)
```

Comment:
How about a test for a pre-existing langpack, and someone uploads a new one, but there's a signing error. To verify that the old one is still served, etc?